### PR TITLE
Cherry-pick: fix metrics for establish connection

### LINKED ISSUE #6197: `mpp_establish_conn` related metrics is broken when enable async grpc server
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

When enable async grpc server, `mpp_establish_conn` related metrics(mpp_establish_conn_duration, mpp_establish_conn_count) is broken because in https://github.com/pingcap/tiflash/blob/731482ec0bf3240622af854db3170b81eff657bb/dbms/src/Flash/FlashService.cpp#L256-L260
it is assumed that when `AsyncFlashService::establishMPPConnectionAsync` finishes, the establish connection also finishes. It is not true for async grpc server.

### LINKED ISSUE #6656: Cherry-pick: fix metrics for establish connection

### PR DESCRIPTION
**Pull request by xzhangxian1008**
### What problem does this PR solve?

Issue Number: close #6197

Problem Summary:

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Flash/EstablishCall.cpp [hunk 1]
```diff
@@ -37,6 +37,11 @@ EstablishCallData::EstablishCallData(AsyncFlashService * service, grpc::ServerCo
 EstablishCallData::~EstablishCallData()
 {
     GET_METRIC(tiflash_object_count, type_count_of_establish_calldata).Decrement();
+    if (stopwatch)
```
Comment (windtalker) [inline dbms/src/Flash/EstablishCall.cpp:40]: also need to update `EstablishConnectionSyncOrAsync` to not update these metrics for async connection.

File: dbms/src/Flash/FlashService.cpp [hunk 2]
```diff
@@ -230,25 +220,42 @@ ::grpc::Status FlashService::establishMPPConnectionSyncOrAsync(::grpc::ServerCon
...
-    if (!security_config.checkGrpcContext(grpc_context))
+    auto check_result = checkGrpcContext(grpc_context);
+    if (!check_result.ok())
+        return check_result;
+
+    if (calldata)
     {
-        return returnStatus(calldata, grpc::Status(grpc::PERMISSION_DENIED, tls_err_msg));
+        GET_METRIC(tiflash_thread_count, type_active_threads_of_establish_mpp).Increment();
+        GET_METRIC(tiflash_thread_count, type_total_threads_of_raw).Increment();
```
Comment (windtalker) [inline dbms/src/Flash/FlashService.cpp:230]: Don't need to update these metric for async connection, instead, should update `tiflash_coprocessor_request_count` and `tiflash_coprocessor_handling_request_count`

File: dbms/src/Flash/FlashService.cpp [hunk 3]
```diff
@@ -230,25 +220,42 @@ ::grpc::Status FlashService::establishMPPConnectionSyncOrAsync(::grpc::ServerCon
...
     Stopwatch watch;
     SCOPE_EXIT({
-        GET_METRIC(tiflash_thread_count, type_total_threads_of_raw).Decrement();
-        GET_METRIC(tiflash_thread_count, type_active_threads_of_establish_mpp).Decrement();
-        GET_METRIC(tiflash_coprocessor_handling_request_count, type_mpp_establish_conn).Decrement();
-        GET_METRIC(tiflash_coprocessor_request_duration_seconds, type_mpp_establish_conn).Observe(watch.elapsedSeconds());
+        if (calldata)
+        {
+            GET_METRIC(tiflash_thread_count, type_total_threads_of_raw).Decrement();
+            GET_METRIC(tiflash_thread_count, type_active_threads_of_establish_mpp).Decrement();
```
Comment (windtalker) [inline dbms/src/Flash/FlashService.cpp:250]: No need to update metrics for async connection here.

File: dbms/src/Flash/EstablishCall.cpp (hunk 1 above)
Comment (windtalker) [reply dbms/src/Flash/EstablishCall.cpp:40]: I think even if `stopwatch` is nullptr, you should update `tiflash_coprocessor_handling_request_count`

Comment (xzhangxian1008) [reply dbms/src/Flash/EstablishCall.cpp:40]: okk

File: dbms/src/Flash/FlashService.cpp [hunk 4]
```diff
@@ -230,25 +220,42 @@ ::grpc::Status FlashService::establishMPPConnectionSyncOrAsync(::grpc::ServerCon
...
     Stopwatch watch;
     SCOPE_EXIT({
-        GET_METRIC(tiflash_thread_count, type_total_threads_of_raw).Decrement();
-        GET_METRIC(tiflash_thread_count, type_active_threads_of_establish_mpp).Decrement();
-        GET_METRIC(tiflash_coprocessor_handling_request_count, type_mpp_establish_conn).Decrement();
-        GET_METRIC(tiflash_coprocessor_request_duration_seconds, type_mpp_establish_conn).Observe(watch.elapsedSeconds());
+        if (calldata)
+        {
+            GET_METRIC(tiflash_coprocessor_handling_request_count, type_mpp_establish_conn).Decrement();
+            GET_METRIC(tiflash_coprocessor_request_duration_seconds, type_mpp_establish_conn).Observe(watch.elapsedSeconds());
```
[the remark this answers is not in the document]
Comment (xzhangxian1008) [reply dbms/src/Flash/FlashService.cpp:250]: > y update these metrics here? They shou

okk, I will delete them here

**Review by ywqzzy — APPROVED** [review]

**Pull request merged**